File: server/api/canvas.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

func (c *APIController) GetCanvasOAuth2Url(w http.ResponseWriter, r *http.Request) (int, error) {
	params := url.Values{}

	params.Add("client_id", c.canvas.ClientID)
	params.Add("response_type", "code")
	params.Add("state", "state")
	params.Add("redirect_uri", "redirect_uri")

	url := fmt.Sprintf("%s/login/oauth2/auth?%s", c.canvas.HtmlUrl, params.Encode())

	payload := struct {
		Url string `json:"url"`
	}{
		Url: url,
	}

	if err := json.NewEncoder(w).Encode(payload); err != nil {
		return http.StatusInternalServerError, err
	}

	return http.StatusOK, nil
}

// func (c *APIController) GetCanvasOAuth2(w http.ResponseWriter, r *http.Request) (int, error) {
// 	code := chi.URLParam(r, "code")
// 	if code == "" {
// 		return http.StatusBadRequest, fmt.Errorf("missing code")
// 	}

// 	state := chi.URLParam(r, "state")
// 	if state == "" {
// 		return http.StatusBadRequest, fmt.Errorf("missing state")
// 	}

// 	types := []canvas.CourseEnrollmentType{canvas.StudentCourseEnrollment}

// 	results, code, err := c.canvas.HtmlUrl
// 	if err != nil {
// 		return code, err
// 	}

// 	if err := json.NewEncoder(w).Encode(results); err != nil {
// 		return http.StatusInternalServerError, err
// 	}

// 	return http.StatusOK, nil
// }

File: web/src/api/canvas.ts
import { SupabaseClient } from '@supabase/supabase-js';
import axios from 'axios';
import { CanvasOAuth2 } from '../entities/canvas';

export const getCanvasOAuth2 = async (
  supabase: SupabaseClient,
  code: string,
  state: string
) => {
  try {
    const accessToken = (await supabase.auth.getSession()).data.session
      ?.access_token;

    const { data } = await axios(`/oauth2`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      params: {
        code: code,
        state: state,
      },
    });

    return data as CanvasOAuth2;
  } catch (err) {
    throw err as Error;
  }
};

export const getCanvasOAuth2Url = async (supabase: SupabaseClient) => {
  try {
    const accessToken = (await supabase.auth.getSession()).data.session
      ?.access_token;

    const {
      data: { url },
    } = await axios(`/oauth2/url`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    });

    return url as string;
  } catch (err) {
    throw err as Error;
  }
};
